### ISSUE TITLE
Game lagging

### ISSUE THREAD

**Issue by rb431** [Misunderstood-Wookiee/Mules-and-Warehosues-Extended#120]
Through trial and error, I have determined that this particular add-in is causing me serious lag. I don't use an extensive list of add-in's just these:
Both DLC's
Escape Teleport 1.3
Faction Enhancer (Catchup, Econ AI, Econ Balance, War) 5.3
Inventory Collector 1.13
Mules, Supply and Warehouses Extended 6.32
Sector Patrol 1.05
SirNukes Mod Support APIs 1.86
Smooth Mouse Steering 1.0
Trader Seminars 1.02
Urgent Orders 1.1

So I started noticing that the game would frequently "hang". Sometimes for less than a second, sometimes a second or so but not longer than that. In fact, it's only when I'm actually flying a ship that it's intolerable. It's enough that combat is very poor and not fun at all.

To troubleshoot, I disabled all add-ins and started from my latest save point just vanilla. Smooth as glass. I then started adding the extensions back in and it's clear that Mules is it. Mules off, nice and smooth. Mules on, momentary game pauses very frequently. 

I've got a dozen or so M1 mules for my PHQ station, dozens of M2 mules for my various stations, and 20-30 M4 mules. I'm guessing this is some sort of calculation issue. I'd really, really hate to have to live without this mod so I'm happy to perform any sort of troubleshooting that would be helpful.

I'm on a modern motherboard, 32GB of system RAM, and a good Nvidia card. I've watched the system using Performance Monitor and my system utilization isn't spiking while this is going on.

**Comment by bhayden53** [comment]
There is a patch for the station mule that fixes some potential lag situations; @Misunderstood-Wookiee was going to deploy that a couple of days ago but I haven't seen it yet.

By M2, do you mean distribution mules? I rarely use them and generally advise users to avoid it due to logic headaches; I am not sure exactly what it's potential performance issues might be.

On the supply mule, the best way to reduce lag is to reduce the trade space; that means making sure the max jumps is only as large as you need it to be, and if you have no intention for them to ever trade with the AI, unchecking that box. Even if you have trade rules applied, if you have the box checked to trade with AI, they will still have to pull in all the possible trades, and then rule them out one-by-one with the trade rule logic. If you uncheck the box, the trades never make it into the logic checking at all.

**Comment by bhayden53** [comment]
another note on supply mule; by default, the mule starts with max jumps of usually 4/4 or 6/6 depending on pilot skill. As the pilot levels, that jump range stays pinned at max; so at some point after you start the mule, you usually want to go back and reduce the jumps below his maximum so it will stop growing every time the pilot levels.

**Comment by Misunderstood-Wookiee** [comment]
Oh crap, I forgot, rl drama ensued. I'm almost back at the  computer ill get this patch sorted in the next hr.

**Comment by rb431** [comment]
So generally, I provide at least 2 M2 Station Mules to my production stations to move goods back and forth to my Warehouse and retrieve food and meds. Bigger stations, like Engine Parts, get more M2's to service the station and keep the goods flowing to the warehouse. I use some M1's to sell goods around space to the AI stations for cash. I use a lot of M4's which are great for starting up a new station build and just general busywork. So one of the other mods I use is "trader seminars" which makes the skill books easily available to skill up pilots. Most of my M4's have 16 or more jumps available to trade so perhaps I sabotaged myself by making too much of the universe available to too many ships?

**Comment by rb431** [comment]
> 
> 
> Oh crap, I forgot, rl drama ensued. I'm almost back at the computer ill get this patch sorted in the next hr.

I'd just LOVE not to have to redo all my M4's to reduce their coverage if this takes care of things. 👍

**Comment by bhayden53** [comment]
This patch will not reduce the M4 jump range for you, unfortunately. That's just something there's not currently a great workaround for.

The travel mule max jumps is also good to keep relatively small. 

I'd say your biggest performance gains would be:
- keep your travel mule jumps smaller
- reduce your supply mule jumps
- remove AI trades from supply mules you only want trading in player goods

**Comment by bhayden53** [comment]
Something worth mentioning, in my own personal save, the biggest impact on my own stuttering and framerate in my game was when I installed the faction enhancer mods. Possibly it's due to the number of ships increasing with the logic there, or the jobs changing, but definitely after I added that to my save, I noticed some stuttering.

I am not solely blaming that mod; any mods in X4 could be causing lag, and generally, the more you have, the more likely you're going to get frames packed full of poorly optimized modder code.

**Comment by rb431** [comment]
> 
> 
> This patch will not reduce the M4 jump range for you, unfortunately. That's just something there's not currently a great workaround for.
> 
> The travel mule max jumps is also good to keep relatively small.
> 
> I'd say your biggest performance gains would be:
> 
>     * keep your travel mule jumps smaller
> 
>     * reduce your supply mule jumps
> 
>     * remove AI trades from supply mules you only want trading in player goods

Yeah, I discovered that I could reduce the range without having to remove and re-add the assignment so I did that on all my M4's. I also removed the AI stations from about half of them. I maxed them out a 8 jumps now. Some of them were in the 20's or more jumps.

**Comment by rb431** [comment]
> 
> 
> Something worth mentioning, in my own personal save, the biggest impact on my own stuttering and framerate in my game was when I installed the faction enhancer mods. Possibly it's due to the number of ships increasing with the logic there, or the jobs changing, but definitely after I added that to my save, I noticed some stuttering.
> 
> I am not solely blaming that mod; any mods in X4 could be causing lag, and generally, the more you have, the more likely you're going to get frames packed full of poorly optimized modder code.

If I can't get he stuttering under control otherwise, I will ditch those in favor of Mules. I find that oh so much more useful.

**Comment by Misunderstood-Wookiee** [comment]
To quote this, I haven't used those sets of mods since 4.0 as they largely
just add a couple more stations but most of the essential use case is now
fixed by egosoft

**Comment by bhayden53** [comment]
In order to manage your expectations, I would say that even a totally vanilla x4 is likely to stutter. Another cause of stuttering is simply the more satellites you have.

**Comment by rb431** [comment]
> 
> 
> In order to manage your expectations, I would say that even a totally vanilla x4 is likely to stutter. Another cause of stuttering is simply the more satellites you have.

Yeah, one of the things I do once the Adv. Sat's are not too expensive for my wallet is to have my mapping ships go around and take down my original ones and put back fewer advanced. Now that I have trade subscriptions I can really take down any that I have just for trading purposes since I don't need them to update orders anymore. I like to have them at the jump gates and any conflict points just to keep an eye on things.

**Comment by bhayden53** [comment]
I find the sector satellites mod to be absolutely essential for reducing satellite-based lag. You can place an advanced sat out in the middle of nowhere in the sector and it will keep your trade subscriptions for the whole sector updated, and even auto-discover new stations as they're built. 

It's a bit cheat-y, compared to vanilla, but it drastically reduces the amount of micro required, and also the number of satellites.

**Comment by rb431** [comment]
Btw, let me ask some advice. I've basically setup all my production in the NF VI sector aside from things like graphene/antimatter cell/Teladinium, which are in the resource sectors. I have a big warehouse in Pious Mists II on the super highway and I'd like to setup another warehouse up "north" to supply shipbuilding stations. What do you think would be the most effective setup to keep the warehouses full? M2's from the production stations to the PM II warehouse, and then M2's from the PM II warehouse to the norther ware house to keep it full?

**Comment by bhayden53** [comment]
In previous versions of mules the station mule wasn't allowed to trade between trade stations, but this incoming change actually removes that restriction. Station mules could certainly do the job of moving things between trade stations, although I am not sure how well they'll actually do the job. You may have to carefully monitor stock levels to make sure the station mule doesn't totally empty you out somewhere. There is also some amount of randomness to exactly which ware he will pick on any given run.

For this, I think I would probably assign supply mules to the distant tradestations with enough jumps to reach your primary areas. Turning AI trades off, of course :)

This was actually the original purpose of the distribution mule, but in general I have found the logic to be poor and have not had a chance to significantly improve it, as it's pretty complex and breaking it completely is likely.

**Comment by Misunderstood-Wookiee** [comment]
Patch Deployed as of today.

**Comment by Misunderstood-Wookiee** [comment]
So with faction fixes mod from what  I understand using the mod today you wind up with a couple of extra stations to help support fleets which actually are pretty laggy for the game ship production that is, and I think it messes with fleet sizes and amount of ships in the game as a whole which is the real impact in frame rate it as that taxes the CPU more than vanilla size ship counts. 

Last time I tested it the fps difference was around Bet 3.20 or something and I think it was bout 4-8 fps on stations.
anything which removes asteroids from the players perspective helps a TON around those regions I actually have my own mod up for that https://steamcommunity.com/sharedfiles/filedetails/?id=2123736399

Largely because I think the small roids are useless and are just screwing with collision issues in sector players are in along with just drain fps, (visually seen asteroids are not what the AI use to calculate mining so less of that makes no difference)

**Comment by baldamundo** [comment]
> 
> 
> There is a patch for the station mule that fixes some potential lag situations; @Misunderstood-Wookiee was going to deploy that a couple of days ago but I haven't seen it yet.
> 
> By M2, do you mean distribution mules? I rarely use them and generally advise users to avoid it due to logic headaches; I am not sure exactly what it's potential performance issues might be.
> 

Sorry, are you talking about the "M2-Station Mule" or the "M3-Distribute Mule"? I've been using a decent number of both and I've been having exactly the same periodic lag-spikes that the OP describes. I've also been having what seem to be some quite serious logic issues with the M2-Station Mules. 

I've been banging my head against the wall trying to work out if I've set it up wrong, but I've eventually had some behaviour that is clearly not WAD. Maybe I ought to post a separate thread, but as I say, it seems to relate to some of what's mentioned above.

Issues I've had with M2-Station Mules:

1. when set to move goods one-way from a factory to a warehouse, sometimes they seem to ignore what should be valid trades
2. when in-system trade is enabled, sometimes the mules will buy goods from the warehouse and then immediately sell those goods directly back to the warehouse


Although sorry, now I've been experimenting with it and may have found a more fundamental issue - although i don't know if it's the mod or vanilla: station trade offers aren't working correctly for fleet subordinates. If I try to trade with one of my stations with an independent ship or fleet leader, it correctly displays the station buy offer. If I try to trade with a subordinate fleet member, the station will display as only wanting to buy a much lower quantity of goods.  Seems to affect all my ships and all my stations, but not AI stations? And seems to affect manual trades as well as supply mule trades.

Anyone have any idea what that's about? It's infuriating, because setting up this many traders one at a time without fleet mimic is going to be a nightmare.

**Comment by bhayden53** [comment]
The last bit of what you said sounds like a funding issue. It is using the account of the station it is a subordinate of. If a station manager doesn't have the full funding they're requesting, then you'll see things like, one ship will reserve a trade to bring the ware, and then subsequent ships will no longer even see the ware in their trade offers. If you go to logical overview or the right click menu, it will show that the station has demand for it, so it seems like everything should be good; but the actual function in the game that returns trade offers just won't return anything, or will return very few. My best guess is that this is the way the engine is protecting funding to buy other wares; once there's already a reservation for a ware, if there's not a ton of funding, they'll save their money for other wares until that trade completes. It's totally a guess because all of that happens deep in the engine, away from the scripts that are moddable. But this kind of thing usually happens if you don't give the station all the funding it's asking for.

That is actually also probably the cause to your 1) above, but you'll have to be way more specific if you want me to look into something.

for 2), I have seen really dumb behavior out of the distribution mule, but I have never seen a station mule do anything like this, so if you catch this happening, please provide as much detail as possible.

**Comment by rb431** [comment]
I have definitely seen funding as a big issue. If the station manager doesn't have what they think are enough credits, then they just seem to stop doing things. I setup a station with only container storage to use a a warehouse. No production modules at all. I then setup M2 mules at all the production stations pointing to the warehouse as a target. At first nothing moved at all. I had zeroed out the warehouse station's account since I had set it for only internal trades and not to buy things from AI vendors. The estimated budget was something like 54 million credits and once I accepted that estimate then it broke the log jam and goods started moving the way they should. Production stations stock the warehouse and take back food and meds from the warehouse back to the production station. Then I use a lot of M4's with no AI purchases to do most of the distribution to stations like my ship construction stations. 

So one think I learned is that reservations count as if the storage had the item in stock. I started removing all the production modules that I had built at PHQ, so its build storage was just full of claytronics and hull parts. I took and incarcatura and ordered it to do round trip after round trip to my warehouse to empty the build storage at PHQ and when I did that all my hull part deliveries from my production station stopped. I finally figured out that it was because I had setup like 20k in reservations from the build storage to the warehouse so the warehouse stopped taking from the hull part production station. I had to remove most of those trades from the order queue of that one ship to get it moving again.

**Comment by rb431** [comment]
So one problem I'm facing is that I have to move my warehouse. I'm trying to finish the Paranid story line where they want you to build the station with the three throne modules. My problem is that I had purchased a plot in Pious Mist II for my warehouse and now that's where the mission wants me to build the damn station. So I bought another plot nearby and I'm building a replacement warehouse. My challenge is that I have dozens of M2's setup for my production stations with that first warehouse as a target and I can't change the target station without removing the default order and setting the whole thing back up. I can't just click and change that one item. I think I've found where I need to make changes in the save file, but it's hard working on a 1.2GB text file. Is there any way to make it so that I could modify the target station without removing the default M2 order??

**Comment by bhayden53** [comment]
there used to be a mod that unlocked a ship's orders when it was a subordinate, but I don't think it's been updated in forever and it was UI dependent so it probably doesn't work anymore. 

So unfortunately I don't think there's any good way to mass-change the target station. Just put on some music and push through it, then make sure to save right after you're done!

**Comment by rb431** [comment]
> So unfortunately I don't think there's any good way to mass-change the target station. Just put on some music and push through it, then make sure to save right after you're done!

Nope, not gonna do it through the UI. I'll edit the save game and do it that way. Notepad++ isn't usable at those file sizes, but Ultraedit seems to work.

**Comment by bhayden53** [comment]
Right, well, change my statement to "back up your save before editing it!"

**Comment by rb431** [comment]
My plan is to pause the game and make two separate saves. I can go to town on one and still have a reliable backup in case I pooch it somehow.

**Comment by rb431** [comment]
Worked like a treat! I had to use Ultraedit since the file was so big, but I changed all the Home, targetStation, and sourceStation tags that had the original warehouse ID to the new ID and it worked the first try. It even waited to move the subordinates from the old warehouse to the new until after their queued commands finished. Just a few simple search and replaces. 

I did not try and change all the active trade destinations, I just let them finish as ordered. I figured changing those would have other ramifications I wouldn't understand.
